src/index.js

```diff
@@ -1,67 +1,121 @@
+}
+
+// On success callback
+const success = (fileName, installPath, callback) => {
+  // go-ipfs contents are in 'go-ipfs/', so append that to the path
+  const outputPath = installPath + '/go-ipfs/'
+  process.stdout.write(`Downloaded ${fileName}\n`)
+  process.stdout.write(`Installed go-${fileName.replace('.tar.gz', '').replace('.zip', '').replace(/_/g, ' ')} to ${outputPath}\n`)
+  callback({ file: fileName, dir: outputPath })
+}
+
+// Main function
+function download (version, platform, arch, installPath) {
+  return new Promise((resolve, reject) => {
+    //            Environment Variables         Args        Defaults
+    version     = process.env.TARGET_VERSION || version  || goIpfsVersion
+    platform    = process.env.TARGET_OS      || platform || goenv.GOOS
+    arch        = process.env.TARGET_ARCH    || arch     || goenv.GOARCH
+    installPath = installPath ? path.resolve(installPath) : path.resolve(process.cwd())
+
+    // Make sure we support the requested package
+    try {
+      support.verify(version, platform, arch)
+    } catch (e) {
+      return error(e, reject)
     }
-  })
 
-  process.stdout.write(`downloading ${url}\n`)
-
-  if (isWindows) {
-    fileStream.pipe(
-      unzip.Extract({ path: installPath })
-      .on('close', callback))
-  } else {
-    fileStream.pipe(gunzip())
-    .pipe(
-      tarFS
-        .extract(installPath)
-        .on('finish', callback)
-    )
-  }
+    // Flag for Windows
+    const isWindows = support.isWindows(platform)
+
+    // Create the download url
+    const fileExtension = isWindows ? '.zip' : '.tar.gz'
+    const fileName = 'ipfs_' + version + '_' + platform + '-' + arch + fileExtension
+    const url = distUrl + '/go-ipfs/' + version + '/go-' + fileName
+
+    // Success callback wrapper
+    const done = () => success(fileName, installPath, resolve)
+
+    // Unpack the response stream
+    const unpack = (stream) => {
+      // TODO: handle errors for both cases
+      if (isWindows) {
+        return stream.pipe(
+          unzip
+            .Extract({ path: installPath })
+            .on('close', done)
+        )
+      }
+
+      return stream
+        .pipe(gunzip())
+        .pipe(
+          tarFS
+            .extract(installPath)
+            .on('finish', done)
+        )
+    }
+
+    // Start
+    process.stdout.write(`Downloading ${url}\n`)
+
+    request.get(url, (err, res, body) => {
+      // Handle errors
+      if (res.statusCode !== 200)
+        error(new Error(`${res.statusCode} - ${res.body}`), reject)
+    })
+    .on('response', (res) => {
+      // Unpack only if the request was successful
+      if (res.statusCode !== 200)
+        return 
+
+      unpack(res)
+    })
+  })
 }
 
-function noop () {}
+// Public interface
+Object.assign(download, { Platforms: support.Platforms })
+Object.assign(download, { Versions: support.Versions })
+Object.assign(download, { Archs: support.Archs })
+Object.assign(download, { Download: download })
```

Comment:
I refactored the public api. I still think the uppercased properties are nicer, so keeping them as such. We need to have the download function as our module.export in order to not break the API for users (backwards compatibility).